### DISCUSSION TITLE
How to implement new coupled intercalation kinetics

### DISCUSSION THREAD

**Discussion by 1836005678** [pybamm-team/PyBaMM#2978] [Q&A]
Hello, everyone.
I want to implement a new intercalation kinetics, just like equation (5) below, coming from reference [1] and  [2].
![image](https://github.com/pybamm-team/PyBaMM/assets/57339960/8d256f89-d960-4e12-b5c8-8705fa9bedb3)
The implemention of equation (5) needs the implementation of Negative/Postive particle surface concentration [mol.m-3], Average negative/postive particle concentration [mol.m-3], Negative/Postive electrolyte concentration [mol.m-3], X-averaged negative/postive electrolyte concentration [mol.m-3] and maybe other variables in advance.
However, in current PyBaMM dfn models, these variables are implemented in {self.set_particle_submodel() -> pybamm.particle.FickianDiffusion} and {self.set_electrolyte_concentration_submodel() -> pybamm.electrolyte_diffusion.Full} after the implementation of intercalation kinetics in {self.set_intercalation_kinetics_submodel(). -> pybamm.kinetics.SymmetricButlerVolmer).
![image](https://github.com/pybamm-team/PyBaMM/assets/57339960/f09989fd-e326-42e8-a7cf-8a9ee4318150)
This means, if we just rebuild intercalation kinetics code, bugs like 'variables missing' will occur.
![image](https://github.com/pybamm-team/PyBaMM/assets/57339960/f1850725-ad55-445f-bbac-fa39524451f8)
To implement equation (5), we may need to rebuild electrolyte concentration submodel, particle submodel and intercalation kinetics submodel together at once. However, I cannot complete this task on my own. So i am looking for help. Here is my rebuilt intercalation kinetics codes for reference.

class SymmetricButlerVolmer(BaseKinetics):

    def __init__(self, param, domain, reaction, options, phase="primary"):
        super().__init__(param, domain, reaction, options, phase)

    def _get_kinetics(self, j0, ne, eta_r, T, u):
        Feta_RT = self.param.F * eta_r / (self.param.R * T)

        domain, Domain = self.domain_Domain
        phase_name = self.phase_name       

        c1m_surface = self.variables[f"{Domain} {phase_name}particle surface concentration [mol.m-3]"]
        c1m_average = self.variables[f"Average {domain} {phase_name}particle concentration [mol.m-3]"]

        if domain == "negative":
            c1m_maximum = 14540
        elif domain == "positive":
            c1m_maximum = 15320
        
        c2 = self.variables[f"{Domain} {phase_name}electrolyte concentration [mol.m-3]"]
        c2_average = self.variables[f"X-averaged {domain} {phase_name}electrolyte concentration [mol.m-3]"]

        jm = j0 * ((c1m_surface * pybamm.exp(ne * 0.5 * Feta_RT) / c1m_average) - ((c1m_maximum - c1m_surface) * c2) * pybamm.exp(- ne * 0.5 * Feta_RT) / ((c1m_maximum - c1m_average) * c2_average))

        return jm

[1] Chayambuka K, Mulder G, Danilov D L, et al. Physics-based modeling of sodium-ion batteries part II. Model and validation[J]. Electrochimica Acta, 2022, 404: 139764.
[2] Danilov D, Niessen R A H, Notten P H L. Modeling all-solid-state Li-ion batteries[J]. Journal of the Electrochemical Society, 2010, 158(3): A215.

**Comment by brosaplanella** [comment]
I think implementing this might be easier than you suggest. I am not too familiar with the various versions of Butler Volmer, but that looks to me as an asymmetric version of Butler Volmer (see [here](https://github.com/pybamm-team/PyBaMM/blob/d90a56c5bc9d8d81c9378c423f7927aa5ec1ed30/pybamm/models/submodels/interface/kinetics/butler_volmer.py#LL39C9-L39C9)). Currently, our asymmetric version takes the same `j0` for both reactions, but it should be fairly easy to change it to take a `j0_ox` and a `j0_red`. 

I am not sure if we should keep the functionality of passing a single `j0` only (not sure if it breaks some models) so it would be good to hear what @tinosulzer thinks.

**Comment by valentinsulzer** [comment]
Don't worry about the order, we try the submodels in a loop until all the variables have been found. 

I think we should have a version of asymmetric Butler-Volmer with one `j0`, and one with `j0_ox` and `j0_red`. Not sure what they should be called.

**Comment by 1836005678** [comment]
I find that the values of those four variables (Negative/Postive particle surface concentration [mol.m-3], Average negative/postive particle concentration [mol.m-3], Negative/Postive electrolyte concentration [mol.m-3], X-averaged negative/postive electrolyte concentration [mol.m-3]) can be got in class BaseKinetics (implemented in base_kinetics.py), which is the superclass of classes in butler_volmer.py. Therefore, we can pass values from class BaseKinetics to class SymmetricButlerVolmer to implement the new coupled intercalation kinetics and even other intercalation kinetics. Solution has been found.
